cratetorrent/src/disk/io/file.rs

```diff
@@ -166,4 +227,100 @@ impl TorrentFile {
 
         Ok(iovecs)
     }
+
+    pub fn read<'a>(&self, file_slice: FileSlice, blocks: &'a mut [u8]) -> Result<&'a mut [u8], ReadError> {
+        let len = file_slice.len as usize;
+
+        read_exact_at(&self.handle, &mut blocks[..len], file_slice.offset)
+            .map_err(|err| {
+                log::warn!("File {:?} read error: {}", self.info.path, err);
+                err
+            })?;
+
+        Ok(&mut blocks[len..])
+    }
+}
+
+#[cfg(target_os = "linux")]
+fn convert_nix_error(err: nix::Error) -> std::io::Error {
+    match err {
+        nix::Error::Sys(errno) => std::io::Error::from_raw_os_error(errno as i32),
+
+        // preadv/pwrite never return them, but better safe than sorry
+        nix::Error::InvalidPath |
+        nix::Error::InvalidUtf8 |
+        nix::Error::UnsupportedOperation => {
+            log::warn!("Unexpected nix::Error kind ({}), falling back to last_os_error", err);
+            std::io::Error::last_os_error()
+        },
+    }
+}
+
+// Here's where the most effective implementation of `write_at/read_at` is chosen.
+// The trick is, linux-specific APIs do not modify the "seek cursor", while Windows' and others' do.
+// This discrepancy doesn't matter because nothing in this crate relies on this cursor
+// (the read/write offset is always provided explicitly).
+//
+// FIXME: On the other hand, if `TorrentInfo.handle` will ever become exposed in our public API,
```

Comment:
I think this is fine because there are no plans to expose the file handle via the API. While there are plans to make the storage backend generic to be configurable, I don't think we'll want to expose individual torrent file handles themselves.

There might be an unrelated potential issue with this approach, however, see next comment.

There will be more thorough stats reporting for individual files later, but I think that is orthogonal to this.